[GEOS-10452] Fixed Role Services for LDAP/Active Directory.

escheper opened the pull request:
[![GEOS-10452](https://badgen.net/badge/JIRA/GEOS-10452/0052CC)](https://osgeo-org.atlassian.net/browse/GEOS-10452)<!-- PR-BADGE: PLEASE DO NOT REMOVE THIS COMMENT -->

This PR fixes a Role Services for LDAP/Active Directory issue [GEOS-10452].

After release 2.15.x in some cases using a LDAP Active Directory "Role Service" does not work anymore.

Tests shows when using an user defined with a CN with comma's (for example: <surname>, <firstname>), getting roles went well until release 2.15.x but after release 2.15.x getting roles results to none.

Getting roles for users etc. of a "Role Service" the GeoServer code uses a lowlevel Spring method: <ldap template>.search(...). In releases until 2.15.x when an user CN is used, the comma is "escaped" with "\". While debugging the currently master (20220619) tests shows that instead of "\" the token "\5C" or "\\" should be used when using the Spring search method. After changing the code roles of all Active Directory users are correctly retrieved.

For this fix this issue the method "escapeSearchString(...)" is added to "LDAPUtils.java".

Also an UnitTest is added.

**Testcase for the PR** 

To test the LDAP Active Directory functionality we have a public AD server available with some users defined. See below some information.

CAUSION: The Azure LDAP server IP address is not static and may be change in the future.

<pre>
LDAP Server:
  ldap://40.113.110.42/dc=geoad,dc=local
Users:
  geouser1
    CN=geouser1,CN=Users,DC=geoad,DC=local
    memberOf:
      CN=Algemeen,CN=Users,DC=geoad,DC=local
    sAMAccountName=geouser1
  geouser3
    CN=geouser3,OU=Users,OU=Accounts,OU=Aris,DC=geoad,DC=local
    memberOf:
      CN=Kadaster,CN=Users,DC=geoad,DC=local
    sAMAccountName=geouser3
  geouser9
    CN=User9, Geo,CN=Users,DC=geoad,DC=local
    memberOf:
      CN=Kadaster,CN=Users,DC=geoad,DC=local
      CN=Algemeen,CN=Users,DC=geoad,DC=local
    sAMAccountName=geouser9
</pre>

Follow the steps below to test this PR.

CAUSION: The specified ip-address may be ch

1) Start GeoServer and login as Admin

2) Add an Authentication Provider
<pre>
Security | Authentication
	Authentication Providers
		Add new
			Click "LDAP"

Name:                                  ad-ldap-test
Server URL:                            ldap://40.113.110.42/dc=geoad,dc=local
TLS:                                   OFF
User lookup pattern:            
Filter used to lookup user:            (sAMAccountName={1})
Format used for user login name:       {0}@geoad.local
Use LDAP groups for authorization:     ON
Bind user before searching for groups: ON
Group search base:                     CN=Users
Group search filter:                   member={0}
Group to use as ADMIN:
Group to use as GROUP_ADMIN:
Enable Hierarchical groups search:     OFF

Save
</pre>

3) Test the Authentication Provider connection

<pre>
Security | Authentication
	Authentication Providers
		Click "ad-ldap-test"

Username:     geouser3
Password:      geoww123
Click "Test Connection"
The message "Connection Successful" will be shown.

Username:     geouser9
Password:     geoww123
Click "Test Connection"
The message "Connection Successful" will be shown.
</pre>

4) Activate Authentication Provider
<pre>
Security | Authentication
	Provider Chain
		Set selected providers to:
			default
			ad-ldap-test
Save
</pre>

5) Add Role Provider
<pre>
Security | Users, Groups, Roles
	Role Services
		Add new
			Click "LDAP"
Name:                                  ldaprs-test
Server URL:                            ldap://40.113.110.42/dc=geoad,dc=local
TLS:                                   OFF
Group search base:                     CN=Users
Group user membership search filter:  
All groups search filter:             
Filter used to lookup user:           
Authenticate to extract roles:         ON
Username:                              geouser1@geoad.local
Password:                              geoww123
Enable Hierarchical groups search:     OFF

Save
</pre>

6) Check Roles
<pre>
Security | Users, Groups, Roles
	Roles
		Click "ldaprs-test"
A list of roles will be shown containing the role "ROLE_KADASTER".
</pre>

7) Protect data layer
<pre>
Security | Data
	Data Security
		Add new rule
Workspace:           sf
Layer and groups:    roads
Acsess mode:         Read

Set Selected Roles to:
	ROLE_KADASTER

Save
</pre>

8) Set Catalog Mode
<pre>
Security | Data
	Catalog Mode
		CHALLENGE

Save
</pre>

9) Copy the OpenLayers URL of the Spearfish roads layer
<pre>
Data | Layer Preview
	Spearfish roads
Copy the OpenLayers URL.
</pre>

10) Test Roads with an unauthorized user
<pre>
Open a new Private Window (use Shft+Ctrl+P in FireFox) and paste:
http://192.168.0.57:8080/geoserver/sf/wms?service=WMS&version=1.1.0&request=GetMap&layers=sf%3Aroads&bbox=589434.8564686741%2C4914006.337837095%2C609527.2102150217%2C4928063.398014731&width=768&height=537&srs=EPSG%3A26713&styles=&format=application/openlayers
Login:
	geouser1
	geoww123
You will be get the message "HTTP ERROR 403 Forbidden".
</pre>

11) Test Roads with an authorized user
<pre>
Open a new Private Window (use Shft+Ctrl+P in FireFox) and paste:
http://192.168.0.57:8080/geoserver/sf/wms?service=WMS&version=1.1.0&request=GetMap&layers=sf%3Aroads&bbox=589434.8564686741%2C4914006.337837095%2C609527.2102150217%2C4928063.398014731&width=768&height=537&srs=EPSG%3A26713&styles=&format=application/openlayers
Login:
	geouser9
	geoww123
A map will be shown.
</pre>



<!-- Please help our volunteers reviewing this PR by completing the following items. 
Ask in a comment if you have troubles with any of them. -->

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [x] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

<!--Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.-->

mprins commented:
I've restored the mandatory PR checklist.
Please fix formatting problems

aaime commented on src/security/ldap/src/main/java/org/geoserver/security/ldap/BindingLdapAuthoritiesPopulator.java:
```diff
@@ -209,10 +209,12 @@ public Set<GrantedAuthority> getGroupMembershipRoles(
                     final String formattedFilter =
                             MessageFormat.format(groupSearchFilter, userDn, username);
+                    // Replace "\" with "\5C" for search.
+                    String formattedFilterNew = formattedFilter.replace("\\", "\\5C"); 
```
The \\5C replacement surprised me at first, but it seems there are even more that might have to be considered:
* https://www.ibm.com/docs/en/db2-data-mgr-console/3.1.x?topic=ldap-requirements-dn-value
* https://ldap.com/ldap-dns-and-rdns/

If that's the case, in addition to adding a test, could you please factor out a utility method doing all the necessary escapes, and then call it from all the places where you did the direct "\\5C" replacement?

escheper replied:
Hopefully I can quickly realize your proposed addition and change.

escheper commented:
As requested I added a function to handle needed escapes.
Also as requested I added an UnitTest.
I am aware that the "Mac OS CI" test is not successful.
I don't have access to  Mac pc, so I'm unable to test those pc's.

escheper commented:
Comment to keep this PR alive.

escheper commented:
Comment to keep this PR alive.

escheper commented:
Is there any chance this fix will be included in the master branch? If so, when will that be?

mprins commented on src/security/ldap/src/test/java/org/geoserver/security/ldap/LDAPUtilsTest.java:
```diff
@@ -0,0 +1,19 @@
+/* (c) 2017 Open Source Geospatial Foundation - all rights reserved
```
```suggestion
/* (c) 2023 Open Source Geospatial Foundation - all rights reserved
```

mprins commented on src/security/ldap/src/main/java/org/geoserver/security/ldap/BindingLdapAuthoritiesPopulator.java:
```diff
@@ -209,10 +209,12 @@ public Set<GrantedAuthority> getGroupMembershipRoles(
                     // Get ldap groups in form of Pair<String,String> -> Pair<name,dn>
                     final String formattedFilter =
                             MessageFormat.format(groupSearchFilter, userDn, username);
+                    // Replace the escape token "\" with "\\" to search.
```
this is superfluous as the method is documented, but why could be documented

```suggestion
                    // search requires an escaped string
```

escheper replied:
Thanks for your review. Great.
I'm not a real Java developer and I'm not exactly sure what you mean.

Is your suggestion to replace
   `String formattedFilterNew = LDAPUtils.escapeSearchString(formattedFilter);`
with 
   `formattedFilter = LDAPUtils.escapeSearchString(formattedFilter);`?

Or/and replace all calls to `escapeSearchString` with `searchString.replace("\\", "\\\\")`"?

I will follow your suggestion for the text change.

escheper replied:
And I will follow your suggestion for the comment change.

mprins replied:
Yes, my suggestion is to use `formattedFilter = LDAPUtils.escapeSearchString(formattedFilter);`

escheper replied:
Fine, I shall implement your suggestion one of these days.

mprins commented on src/security/ldap/src/main/java/org/geoserver/security/ldap/BindingLdapAuthoritiesPopulator.java:
```diff
@@ -278,10 +280,12 @@ private void searchNestedGroupMembershipRoles(
                     // Get ldap groups in form of Pair<String,String> -> Pair<name,dn>
                     final String formattedFilter =
                             MessageFormat.format(nestedGroupSearchFilter, groupDn, groupName);
+                    // Replace the escape token "\" with "\\" to search.
```
see above

```suggestion
                    // search requires an escaped string
```

mprins commented on src/security/ldap/src/main/java/org/geoserver/security/ldap/LDAPRoleService.java:
```diff
@@ -354,10 +354,13 @@ public GeoServerRole getGroupAdminRole() {
     public int getRoleCount() throws IOException {
         AtomicInteger count = new AtomicInteger(0);
+
+        // Replace the escape token "\" with "\\" to search.
```
```suggestion
        // search requires an escaped string
```

mprins commented on src/security/ldap/src/main/java/org/geoserver/security/ldap/LDAPUserGroupService.java:
```diff
@@ -170,10 +170,16 @@ protected ContextMapper<Object> addToUsers(SortedSet<GeoServerUser> users) {
     public SortedSet<GeoServerUser> getUsers() {
         final SortedSet<GeoServerUser> users = new TreeSet<>();
 
+        // Replace the escape token "\" with "\\" to search.
```
```suggestion
        // search requires an escaped string
```

mprins commented on src/security/ldap/src/main/java/org/geoserver/security/ldap/LDAPUserGroupService.java:
```diff
@@ -430,21 +436,29 @@ private String getGroupDn(GeoServerUserGroup group) {
     public int getUserCount() {
         AtomicInteger size = new AtomicInteger(0);
+
+        // Replace the escape token "\" with "\\" to search.
```
```suggestion
        // search requires an escaped string
```

mprins commented on src/security/ldap/src/main/java/org/geoserver/security/ldap/LDAPUserGroupService.java:
```diff
@@ -474,12 +488,19 @@ public int getUserCountHavingProperty(String propname) {
     public SortedSet<GeoServerUser> getUsersNotHavingProperty(String propname) {
         final SortedSet<GeoServerUser> users = new TreeSet<>();
 
+        // Replace the escape token "\" with "\\" to search.
```
idem

mprins commented on src/security/ldap/src/main/java/org/geoserver/security/ldap/LDAPUserGroupService.java:
```diff
@@ -488,12 +509,20 @@ public SortedSet<GeoServerUser> getUsersNotHavingProperty(String propname) {
     public int getUserCountNotHavingProperty(String propname) {
         AtomicInteger size = new AtomicInteger(0);
+
+        // Replace the escape token "\" with "\\" to search.
```
idem

mprins commented on src/security/ldap/src/main/java/org/geoserver/security/ldap/LDAPUserGroupService.java:
```diff
@@ -503,12 +532,15 @@ public SortedSet<GeoServerUser> getUsersHavingPropertyValue(String propname, Str
             throws IOException {
         final SortedSet<GeoServerUser> users = new TreeSet<>();
 
+        // Replace the escape token "\" with "\\" to search.
```
idem

mprins reviewed (commented):
not sure creating an extra String for the escaped search argument is worth it, AFAIC it can be inlined being a static method

aaime commented:
@mprins are you in a position to review the functional changes in this PR and approve it? My knowledge of Active Directory is basically nil, while I can have a formal look, I don't actually understand the changes being made.

I'm looking at which PRs can be merged before the big "org.opengis" package rename happens next week, which will create conflicts in almost all PRs (thought maybe not in this one).

escheper commented:
@mprins The mentioned AD/ldap service is a temporary Azure service. It is temporary as we run it based on our free Azure credits from our MSDN account, due to lack of funds...
I just started the Azure VM. This always results in a new IP address. At the moment this is: 40.113.110.42.
I hope you can use this to test my PR.

ianturton commented:
We have a small LDAP server in the docs (https://geoserver.org/acme-ldap/acme-ldap-1.0.jar) that could probably be used for testing
